File: web/src/engine/PixelGrid.ts
import type { CellCoord } from './types'

export class PixelGrid {
  readonly columns: number
  readonly rows: number
  readonly paletteSize: number

  private paletteIndices: Uint8Array
  private painted: Uint8Array
  private _unpaintedCount: number
  private _unpaintedPerColour: Int32Array

  private constructor(
    columns: number,
    rows: number,
    paletteIndices: Uint8Array,
    paletteSize: number,
    painted: Uint8Array,
    unpaintedPerColour: Int32Array
  ) {
    this.columns = columns
    this.rows = rows
    this.paletteIndices = paletteIndices
    this.paletteSize = paletteSize
    this.painted = painted
    this._unpaintedPerColour = unpaintedPerColour

    let count = 0
    for (let i = 0; i < unpaintedPerColour.length; i++) {
      count += unpaintedPerColour[i]
    }
    this._unpaintedCount = count
  }

  static create(columns: number, rows: number, paletteIndices: Uint8Array, paletteSize: number): PixelGrid {
    const totalCells = columns * rows
    const painted = new Uint8Array(totalCells)
    const unpaintedPerColour = new Int32Array(paletteSize)

    for (let i = 0; i < totalCells; i++) {
      const idx = paletteIndices[i]
      if (idx < paletteSize) {
        unpaintedPerColour[idx]++
      }
    }

    return new PixelGrid(columns, rows, paletteIndices, paletteSize, painted, unpaintedPerColour)
  }

  private isInBounds(col: number, row: number): boolean {
    return col >= 0 && col < this.columns && row >= 0 && row < this.rows
  }

  paint(col: number, row: number): boolean {
    if (!this.isInBounds(col, row)) {
      return false
    }

    const flatIdx = row * this.columns + col
    if (this.painted[flatIdx]) {
      return false
    }

    this.painted[flatIdx] = 1
    const paletteIdx = this.paletteIndices[flatIdx]
    if (paletteIdx < this.paletteSize) {
      this._unpaintedPerColour[paletteIdx]--
      this._unpaintedCount--
    }

    return true
  }

  paintCells(coords: CellCoord[]): CellCoord[] {
    const newlyPainted: CellCoord[] = []
    for (const { col, row } of coords) {
      if (this.paint(col, row)) {
        newlyPainted.push({ col, row })
      }
    }
    return newlyPainted
  }

  unpaintCells(coords: CellCoord[]): void {
    for (const { col, row } of coords) {
      if (!this.isInBounds(col, row)) continue

      const flatIdx = row * this.columns + col
      if (this.painted[flatIdx]) {
        this.painted[flatIdx] = 0
        const paletteIdx = this.paletteIndices[flatIdx]
        if (paletteIdx < this.paletteSize) {
          this._unpaintedPerColour[paletteIdx]++
          this._unpaintedCount++
        }
      }
    }
  }

  isComplete(): boolean {
    return this._unpaintedCount === 0
  }

  cell(col: number, row: number): { paletteIndex: number; painted: boolean } {
    if (!this.isInBounds(col, row)) {
      return { paletteIndex: 0, painted: false }
    }

    const flatIdx = row * this.columns + col
    return {
      paletteIndex: this.paletteIndices[flatIdx],
      painted: this.painted[flatIdx] !== 0
    }
  }

  unpaintedIndicesForPaletteIndex(index: number): number[] {
    const result: number[] = []
    for (let i = 0; i < this.paletteIndices.length; i++) {
      if (this.paletteIndices[i] === index && !this.painted[i]) {
        result.push(i)
      }
    }
    return result
  }

  toPaintStateBytes(): Uint8Array {
    const bytes = new Uint8Array(this.painted.length)
    for (let i = 0; i < this.painted.length; i++) {
      const paletteIdx = this.paletteIndices[i]
      bytes[i] = (this.painted[i] ? 0x80 : 0) | (paletteIdx & 0x7f)
    }
    return bytes
  }

  static fromPaintStateBytes(
    bytes: Uint8Array,
    columns: number,
    rows: number,
    paletteSize: number
  ): PixelGrid {
    const totalCells = columns * rows
    const paletteIndices = new Uint8Array(totalCells)
    const painted = new Uint8Array(totalCells)
    const unpaintedPerColour = new Int32Array(paletteSize)

    for (let i = 0; i < Math.min(bytes.length, totalCells); i++) {
      const byte = bytes[i]
      painted[i] = byte & 0x80 ? 1 : 0
      paletteIndices[i] = byte & 0x7f
      if (!painted[i] && paletteIndices[i] < paletteSize) {
        unpaintedPerColour[paletteIndices[i]]++
      }
    }

    return new PixelGrid(columns, rows, paletteIndices, paletteSize, painted, unpaintedPerColour)
  }

  get unpaintedCount(): number {
    return this._unpaintedCount
  }

  unpaintedForColour(index: number): number {
    if (index >= 0 && index < this.paletteSize) {
      return this._unpaintedPerColour[index]
    }
    return 0
  }
}

File: TappyColour/Engine/PixelGrid.swift
import Foundation

// MARK: - RGBA8
struct RGBA8: Equatable, Hashable, Codable {
    let r: UInt8
    let g: UInt8
    let b: UInt8
    let a: UInt8

    init(r: UInt8, g: UInt8, b: UInt8, a: UInt8 = 255) {
        self.r = r
        self.g = g
        self.b = b
        self.a = a
    }
}

// MARK: - GridCell
struct GridCell: Equatable {
    var paletteIndex: UInt8
    var painted: Bool

    init(paletteIndex: UInt8, painted: Bool) {
        self.paletteIndex = paletteIndex
        self.painted = painted
    }
}

// MARK: - PixelGrid
struct PixelGrid: Equatable {
    let columns: Int
    let rows: Int
    private(set) var cells: [GridCell]
    private(set) var unpaintedCount: Int
    private(set) var unpaintedPerColour: [Int]

    // MARK: Initializer
    init(columns: Int, rows: Int, cells: [GridCell]) {
        self.columns = columns
        self.rows = rows
        self.cells = cells

        var unpaintedCount = 0
        var unpaintedPerColour = [Int](repeating: 0, count: 256)

        for cell in cells {
            if !cell.painted {
                unpaintedCount += 1
                unpaintedPerColour[Int(cell.paletteIndex)] += 1
            }
        }

        self.unpaintedCount = unpaintedCount
        self.unpaintedPerColour = unpaintedPerColour
    }

    // MARK: Painting Operations
    mutating func paint(col: Int, row: Int) -> Bool {
        guard let index = flatIndex(col: col, row: row) else { return false }
        guard !cells[index].painted else { return false }

        cells[index].painted = true
        unpaintedCount -= 1
        unpaintedPerColour[Int(cells[index].paletteIndex)] -= 1
        return true
    }

    mutating func paintCells(_ coords: [(col: Int, row: Int)]) -> [(col: Int, row: Int)] {
        var changed: [(col: Int, row: Int)] = []
        for (col, row) in coords {
            if paint(col: col, row: row) {
                changed.append((col, row))
            }
        }
        return changed
    }

    mutating func unpaintCells(_ coords: [(col: Int, row: Int)]) {
        for (col, row) in coords {
            guard let index = flatIndex(col: col, row: row) else { continue }
            guard cells[index].painted else { continue }

            cells[index].painted = false
            unpaintedCount += 1
            unpaintedPerColour[Int(cells[index].paletteIndex)] += 1
        }
    }

    // MARK: State Queries
    var isComplete: Bool {
        unpaintedCount == 0
    }

    func cell(col: Int, row: Int) -> GridCell {
        guard let index = flatIndex(col: col, row: row) else {
            return GridCell(paletteIndex: 0, painted: true)
        }
        return cells[index]
    }

    func unpaintedIndices(forPaletteIndex paletteIndex: UInt8) -> [Int] {
        var indices: [Int] = []
        let targetIndex = Int(paletteIndex)
        for (flatIdx, cell) in cells.enumerated() {
            if Int(cell.paletteIndex) == targetIndex && !cell.painted {
                indices.append(flatIdx)
            }
        }
        return indices
    }

    // MARK: Private Helpers
    private func flatIndex(col: Int, row: Int) -> Int? {
        guard col >= 0, col < columns, row >= 0, row < rows else { return nil }
        return row * columns + col
    }
}
